[serve] Centralize status

zcin opened the pull request:
<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?

Centralize logic that handles status transitions into one location: `handle_transition` in `DeploymentStatusInfo`. This will make it very clear what statuses/status triggers are possible, and which transitions are possible.

## Related issue number

<!-- For example: "Closes #1234" -->

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

edoakes commented on python/ray/serve/_private/common.py:
```diff
@@ -118,17 +118,56 @@ class DeploymentStatusTrigger(str, Enum):
     message: str = ""
 
+    @staticmethod
+    def ranking_order() -> List:
+        return [
+            # Status ranking order is defined in a following fashion:
+            #   1. (Highest) Statuses signalling any failures in the system
```
what's the ranking order used for?

alexeykudinkin commented on python/ray/serve/_private/common.py:
```diff
@@ -118,17 +118,56 @@ class DeploymentStatusTrigger(str, Enum):
 
 
+# Internal Enum used to manage deployment state machine
+class DeploymentStatusTransition(str, Enum):
+    HEALTHY = "HEALTHY"
+    CONFIG_UPDATE = "CONFIG_UPDATE"
+    AUTOSCALE_UP = "AUTOSCALE_UP"
+    AUTOSCALE_DOWN = "AUTOSCALE_DOWN"
+    MANUALLY_INCREASE_NUM_REPLICAS = "MANUALLY_INCREASE_NUM_REPLICAS"
+    MANUALLY_DECREASE_NUM_REPLICAS = "MANUALLY_DECREASE_NUM_REPLICAS"
+    REPLICA_STARTUP_FAILED = "REPLICA_STARTUP_FAILED"
+    HEALTH_CHECK_FAILED = "HEALTH_CHECK_FAILED"
+    INTERNAL_ERROR = "INTERNAL_ERROR"
+    DELETE = "DELETE"
```
So transition is a tuple of `(From State, To State, Trigger)`. 

Occasionally it's trivialized to just refer to a trigger (in case it identifies it uniquely), but i'd still suggest we rather call it trigger (to disambiguate w/ existing one, we can call it `_DeploymentStatusInternalTrigger`)

alexeykudinkin commented on python/ray/serve/_private/common.py:
```diff
@@ -141,6 +180,244 @@ def update(
+            )
+
+        # If deployment is being deleted, set status immediately and return
+        if transition == DeploymentStatusTransition.DELETE:
```
In cases when branches are mutex, let's specify those as `elif` to make it clear (reduces cognitive load of needing to keep multiple control-flows in memory when reading)

alexeykudinkin commented on python/ray/serve/_private/common.py:
```diff
@@ -141,6 +180,244 @@ def update(
+            if transition == DeploymentStatusTransition.AUTOSCALE_UP:
+                # The deployment should only transition to UPSCALING if
+                # it's within the autoscaling bounds
+                if is_within_autoscaling_bounds:
```
nit: Let's just abstract this as `should_autoscale`

alexeykudinkin commented on python/ray/serve/_private/common.py:
```diff
@@ -141,6 +180,244 @@ def update(
+    def transition(
+        self,
+        transition: DeploymentStatusTransition,
+        message: str = "",
```
`message` should actually be dependent on the target state, so it's weird that we pass it in from the outside (where we don't know what the end state would be)

alexeykudinkin replied:
Let's call it `message_override` instead to make it more clear that this is only used to override the message when necessary

zcin replied:
The target state is actually only available on the "outside"; `DeploymentStatusInfo` doesn't have access (for example) to the current number of replicas, or what error happened, etc. I looked into renaming it to message_override, but realized that it's not overriding anything; this function doesn't actually make any decisions to set the message, the message is coming from "outside" 100% of the time.

GeneDer commented on python/ray/serve/_private/common.py:
```diff
@@ -141,6 +199,251 @@ def update(
+            )
+
+        # If deployment is being deleted, set status immediately and return
+        elif trigger == DeploymentStatusInternalTrigger.DELETE:
```
Since all those are early returns, no need to chain the if-else statements together. Just a personal preference, would recommend to just use `if` here instead of `elif` so all the statements are independent. Same goes to all the ones below.

GeneDer commented on python/ray/serve/_private/common.py:
```diff
@@ -141,6 +199,251 @@ def update(
+            elif trigger == DeploymentStatusInternalTrigger.AUTOSCALE_UP:
+                # The deployment should only transition to UPSCALING if
+                # it's within the autoscaling bounds
+                if should_autoscale:
```
Just wondering if the trigger is autoscale up or down, can `should_autoscale` ever be false?

zcin replied:
Yes, if the deployment isn't within the autoscaling bounds, then `should_autoscale` will be false. That is independent of the trigger being `AUTOSCALING`.

zcin replied:
Gotcha, I understand the problem; to make it less awkward, I've made `_set_target_state` only handle setting the target state, and moved the handling of state transitions into `deploy()` and `autoscale()`. In `autoscale()`, it will break early if the deployment isn't within autoscaling bounds.

GeneDer commented on python/ray/serve/_private/common.py:
```diff
@@ -141,6 +199,251 @@ def update(
+            elif trigger in [
+                DeploymentStatusInternalTrigger.MANUALLY_INCREASE_NUM_REPLICAS,
+                DeploymentStatusInternalTrigger.MANUALLY_DECREASE_NUM_REPLICAS,
+            ]:
```
No big deal, but this can just be a set.

GeneDer commented on python/ray/serve/_private/common.py:
```diff
@@ -141,6 +199,251 @@ def update(
+                    message=message,
+                )
+
+        elif self.status in [DeploymentStatus.UPSCALING, DeploymentStatus.DOWNSCALING]:
```
This can also just be a set 🙃

GeneDer commented on python/ray/serve/_private/common.py:
```diff
@@ -141,6 +199,251 @@ def update(
+        trigger: DeploymentStatusInternalTrigger,
+        message: str = "",
+        should_autoscale: bool = False,
+    ):
```
Add a return type `DeploymentStatusInfo`

GeneDer replied:
Also, would be nice to add some docstring saying this function will create new instance of `DeploymentStatusInfo` or possibly returning the existing one. And be explicit that this will not alter the current `DeploymentStatusInfo`.

GeneDer commented on python/ray/serve/_private/deployment_state.py:
```diff
@@ -1461,29 +1457,31 @@ def _set_target_state(
+                    else DeploymentStatusInternalTrigger.MANUALLY_INCREASE_NUM_REPLICAS,
+                    message=f"Upscaling from {curr_num} to {new_num} replicas.",
+                    should_autoscale=self.should_autoscale()
+                    and self._is_within_autoscaling_bounds(),
```
Would be nice to just define those variables before calling `handle_transition()` to be more readable. There are too much logical complexities stuffed in this statements and making it hard to read 😅

zcin replied:
This is now automatically solved because of https://github.com/ray-project/ray/pull/41429#discussion_r1434590715!

GeneDer commented on python/ray/serve/_private/deployment_state.py:
```diff
@@ -1461,29 +1457,31 @@ def _set_target_state(
+                    else DeploymentStatusInternalTrigger.MANUALLY_DECREASE_NUM_REPLICAS,
+                    message=f"Downscaling from {curr_num} to {new_num} replicas.",
+                    should_autoscale=self.should_autoscale()
+                    and self._is_within_autoscaling_bounds(),
```
Same goes here

GeneDer commented on python/ray/serve/_private/deployment_state.py:
```diff
@@ -1553,16 +1551,11 @@ def deploy(self, deployment_info: DeploymentInfo) -> bool:
-        )
+        # allow_scaling_statuses = (
+        #     self.curr_status_info.status is not DeploymentStatus.UPDATING
+        # )
```
Can be deleted 🙃

GeneDer commented on python/ray/serve/_private/deployment_state.py:
```diff
@@ -2161,7 +2121,7 @@ def _check_and_update_replicas(self):
             if len(pending_allocation) > 0:
                 required, available = pending_allocation[0].resource_requirements()
-                message = (
+                msg = (
```
Personally I would prefer to spell those out especially we are not under the character limits here.

zcin commented on python/ray/serve/_private/application_state.py:
```diff
@@ -454,38 +454,13 @@ def _determine_app_status(self) -> Tuple[ApplicationStatus, str]:
-                )
-
-        if len(unhealthy_deployment_names):
+        ranked = sorted(self.get_deployments_statuses(), key=lambda info: info.rank)
```
No, the statuses are ordered, so this is just picking the highest ranked status that exists in `self.get_deployments_statuses`.

zcin commented on python/ray/serve/_private/common.py:
```diff
@@ -141,6 +199,251 @@ def update(
+        self,
+        trigger: DeploymentStatusInternalTrigger,
+        message: str = "",
+        should_autoscale: bool = False,
```
`should_autoscale` is only actually used in one place in the deployment state manager, so I think it makes sense to set the default to False so we don't have to specify it in every other location. What do you think is concerning if we have a default value here?

zcin commented:
@ray-project/ray-serve Hi folks, I've addressed the comments that have been left so far (thanks for those who reviewed!), could folks help take another look?

GeneDer commented on python/ray/serve/_private/common.py:
```diff
@@ -141,6 +200,252 @@ def update(
+            trigger: A (internal) trigger that determines the state
+                transition.
+            message: The message to set in status info.
+            should_autoscale: Whether the state should be allowed to
```
We can drop this `should_autoscale` now :) (and in the args)

GeneDer commented on python/ray/serve/_private/common.py:
```diff
@@ -118,17 +118,76 @@ class DeploymentStatusTrigger(str, Enum):
+                return i
+            elif len(s) == 2 and s == (self.status, self.status_trigger):
+                return i
+
```
Also just want to note the way it's currently implemented, there is a chance this returns None where none of the `ranking_order` are matched. We might want to `return len(DeploymentStatusInfo.ranking_order())` in the end so the upstream comparison won't error out.

zcin replied:
This function is supposed to cover (i.e. return a rank) for all possible deployment statuses. If a deployment doesn't match any of them, that should be a bug; so adding a failsafe like you suggested might cause this bug to go unnoticed (e.g. if we added a new `DeploymentStatus` and forgot to add it to the ranked list). I personally prefer relying on our tests to catch this case, but not sure what the best approach here is.

GeneDer replied:
Yep, totally agreed. That's the other thing I was thinking as well, but thought it's easier to maintain to just have a default catch all. Right now it's already leaking for example when `DeploymentStatus` is `UPSCALING` and `DeploymentStatusTrigger` is `UNSPECIFIED`. Adding a test enumerate all combinations of `DeploymentStatus` and `DeploymentStatusTrigger` and ensure there is a rank sounds good to me👍

GeneDer commented on python/ray/serve/_private/deployment_state.py:
```diff
@@ -1551,18 +1519,29 @@ def deploy(self, deployment_info: DeploymentInfo) -> bool:
+
+            if new_num > old_num:
+                self._curr_status_info = self._curr_status_info.handle_transition(
+                    DeploymentStatusInternalTrigger.MANUALLY_INCREASE_NUM_REPLICAS,
```
Nitpick (non-blocker): Would be nice to use kwargs here. Same goes to few other places below where it's a mix of positional args and kwargs.

GeneDer reviewed (commented):
Mostly LGTM, just wondering if we should change how the rank is calculated.

edoakes commented on python/ray/serve/_private/application_state.py:
```diff
@@ -454,38 +454,13 @@ def _determine_app_status(self) -> Tuple[ApplicationStatus, str]:
-                )
-
-        if len(unhealthy_deployment_names):
+        choice_status = min(self.get_deployments_statuses(), key=lambda info: info.rank)
```
nit: more self-describing
```suggestion
        lowest_rank_status = min(self.get_deployments_statuses(), key=lambda info: info.rank)
```

edoakes replied:
A comment helping to describe the purpose of this line would also be helpful.

It may also be preferable to structure this as a static method on the `DeploymentStatus` class so the rank definitions & logic are centralized there.

edoakes commented on python/ray/serve/_private/common.py:
```diff
@@ -118,17 +118,76 @@ class DeploymentStatusTrigger(str, Enum):
     status_trigger: DeploymentStatusTrigger
     message: str = ""
 
+    @staticmethod
+    def ranking_order() -> List[
+        Tuple[DeploymentStatus, Optional[DeploymentStatusTrigger]]
+    ]:
```
Given this is fully static, why does it need to be a method? This will dynamically re-generate the list each time it's called. Would prefer to make it a class variable or global constant.

zcin replied:
Made it a global constant!

edoakes commented on python/ray/serve/_private/common.py:
```diff
@@ -118,17 +118,75 @@ class DeploymentStatusTrigger(str, Enum):
+        `DeploymentStatusInfo` sub-resources.
+        """
+
+        for i, s in enumerate(DeploymentStatusInfo.ranking_order()):
```
+1, I think you can simply structure the `ranking_order` as a static dictionary

zcin replied:
Thanks for the suggestions guys, restructured it as a global dictionary.

edoakes commented on python/ray/serve/_private/common.py:
```diff
@@ -118,17 +118,76 @@ class DeploymentStatusTrigger(str, Enum):
         return json.dumps(asdict(self), indent=4)
 
-    def update(
+    def _update(
```
nit: I find "update" confusing here because it may be returning a completely new status (or new trigger). Perhaps calling it "transition" is better?

zcin replied:
Makes sense. I'm a little wary of using "transition" though, since it doesn't actually handle any transition logic, and simply functions as an "update the keys of the dictionary" kind of function. I've changed it to `_updated_copy`, how does that sound?

edoakes replied:
sgtm
